README.md
# CO2Meter

Python Module to use co2meters like the 'AirCO2ntrol Mini' from TFA Dostmann with USB ID `04d9:a052`. There are also other modules using the same interface.

This module supports Python 2.7 and 3.x.

## Attribution

- Reverse Engineering of the protocol and initial code done by [Henryk Plötz](https://github.com/henryk). 
- Read all about it at [hackaday](https://hackaday.io/project/5301-reverse-engineering-a-low-cost-usb-co-monitor)
- Code derived from [this article](https://hackaday.io/project/5301-reverse-engineering-a-low-cost-usb-co-monitor/log/17909-all-your-base-are-belong-to-us)
- This repository is a fork of [Michael Nosthoff](https://github.com/heinemml/CO2Meter)'s repo

## Mini CO2 Monitor specification

This is taken from [product page](https://www.tfa-dostmann.de/en/produkt/co2-monitor-airco2ntrol-mini/) and linked Instruction manual.

Traffic LEDs thresholds:

- Green -- CO2 concentration below 800 ppm
- Yellow -- CO2 concentration between 800 ppm and 1200 ppm
- Red -- CO2 concentration above 1200 ppm

Specified operation / measurement ranges:

- Temperature 0°C...50°C
- CO2 concentration -- 0 ... 3000 ppm
  - Resolution of 1 ppm at 0 ... 1000 ppm
  - Resolution of 10 ppm at 1001 ... 3000 ppm

## Install

You can install this lib in standard way, but it is typically enough to just clone the repo and run from it. To allow non root users, you will probably need to install `udev` rules from `docs/99-co2monitor.rules` to `/etc/udev/rules.d/99-co2monitor.rules`. Depending on your system, you may need to run

```bash
udevadm control --reload-rules && udevadm trigger
```

## Usage

```python
from CO2Meter import *
import time
sensor = CO2Meter("/dev/hidraw0")
while True:
    time.sleep(2)
    sensor.get_data()
```

The device writes out one value at a time. So we need to parse some data until we have co2 and temperature. Thus the get_data() method will initially return none or only on value (whichever comes first). When you just need one measurement you should wait some seconds or iterate until you get a full reading. If you just need co2 a call to `get_co2` might speed things up.

### Callback

You can pass a callback to the constructor. It will be called when any of the values is updated. The parameters passed are `sensor` and `value`. `sensor` contains one of these constants:

```python
CO2METER_CO2 = 0x50
CO2METER_TEMP = 0x42
CO2METER_HUM = 0x44
```

### Error handling

In Case the device can't be read anymore (e.g. it was unplugged) the worker thread will end in the background. Afterwards calls to any of the `get_*` functions will throw an `IOError`. You will need to handle any reset, making sure that the device is there etc yourself.


debian/changelog
bobek-co2meter (0.0.1) unstable; urgency=medium

  * Initial release

 -- Antonin Kral <a.kral@bobek.cz>  Thu, 10 Mar 2022 20:39:24 +0100


mqtt.py
#!/bin/env python
from CO2Meter import *
import paho.mqtt.client as mqttClient
import time

def on_connect(client, userdata, flags, rc):
    if rc == 0:
        print("Connected to broker")
        global Connected
        Connected = True
    else:
        print("Connection failed")

Connected = False   #global variable for the state of the connection
Meter = CO2Meter("/dev/hidraw0")

client = mqttClient.Client()
client.on_connect= on_connect
client.connect("10.0.0.250", port=1883)

client.loop_start()

while Connected != True:
    time.sleep(0.1)

try:
    while True:
        measurement = Meter.get_data()
        print(measurement)
        if 'co2' in measurement:
          client.publish("home/livingroom/co2", measurement['co2'])
        if 'temperature' in measurement:
          client.publish("home/livingroom/temperature", measurement['temperature'])
        if 'humidity' in measurement:
          client.publish("home/livingroom/humidity", measurement['humidity'])
        time.sleep(5)
except KeyboardInterrupt:
    client.disconnect()
    client.loop_stop()




CO2Meter.py
import sys
import fcntl
import threading
import weakref

CO2METER_CO2 = 0x50
CO2METER_TEMP = 0x42
CO2METER_HUM = 0x44
HIDIOCSFEATURE_9 = 0xC0094806

def _co2_worker(weak_self):
    while True:
        self = weak_self()
        if self is None:
            break
        self._read_data()

        if not self._running:
            break
        del self


class CO2Meter:

    _key = [0xc4, 0xc6, 0xc0, 0x92, 0x40, 0x23, 0xdc, 0x96]
    _device = ""
    _values = {}
    _file = ""
    _running = True
    _callback = None

    def __init__(self, device="/dev/hidraw0", callback=None):
        self._device = device
        self._callback = callback
        self._file = open(device, "a+b", 0)

        if sys.version_info >= (3,):
            set_report = [0] + self._key
            fcntl.ioctl(self._file, HIDIOCSFEATURE_9, bytearray(set_report))
        else:
            set_report_str = "\x00" + "".join(chr(e) for e in self._key)
            fcntl.ioctl(self._file, HIDIOCSFEATURE_9, set_report_str)

        thread = threading.Thread(target=_co2_worker, args=(weakref.ref(self),))
        thread.daemon = True
        thread.start()


    def _read_data(self):
        try:
            result = self._file.read(8)
            if sys.version_info >= (3,):
                data = list(result)
            else:
                data = list(ord(e) for e in result)

            decrypted = self._decrypt(data)
            if decrypted[4] != 0x0d or (sum(decrypted[:3]) & 0xff) != decrypted[3]:
                print(self._hd(data), " => ", self._hd(decrypted), "Checksum error")
            else:
                operation = decrypted[0]
                val = decrypted[1] << 8 | decrypted[2]
                self._values[operation] = val
                if self._callback is not None:
                    if operation == CO2METER_CO2:
                        self._callback(sensor=operation, value=val)
                    elif operation == CO2METER_TEMP:
                        self._callback(sensor=operation,
                                       value=round(val / 16.0 - 273.1, 1))
                    elif operation == CO2METER_HUM:
                        self._callback(sensor=operation, value=round(val / 100.0, 1))
        except:
            self._running = False


    def _decrypt(self, data):
        cstate = [0x48, 0x74, 0x65, 0x6D, 0x70, 0x39, 0x39, 0x65]
        shuffle = [2, 4, 0, 7, 1, 6, 5, 3]

        phase1 = [0] * 8
        for i, j in enumerate(shuffle):
            phase1[j] = data[i]

        phase2 = [0] * 8
        for i in range(8):
            phase2[i] = phase1[i] ^ self._key[i]

        phase3 = [0] * 8
        for i in range(8):
            phase3[i] = ((phase2[i] >> 3) | (phase2[(i-1+8)%8] << 5)) & 0xff

        ctmp = [0] * 8
        for i in range(8):
            ctmp[i] = ((cstate[i] >> 4) | (cstate[i]<<4)) & 0xff

        out = [0] * 8
        for i in range(8):
            out[i] = (0x100 + phase3[i] - ctmp[i]) & 0xff

        return out


    @staticmethod
    def _hd(data):
        return " ".join("%02X" % e for e in data)


    def get_co2(self):
        if not self._running:
            raise IOError("worker thread couldn't read data")
        result = {}
        if CO2METER_CO2 in self._values:
            result = {'co2': self._values[CO2METER_CO2]}

        return result


    def get_temperature(self):
        if not self._running:
            raise IOError("worker thread couldn't read data")
        result = {}
        if CO2METER_TEMP in self._values:
            result = {'temperature': round((self._values[CO2METER_TEMP]/16.0-273.15),1)}

        return result


    def get_humidity(self): # not implemented by all devices
        if not self._running:
            raise IOError("worker thread couldn't read data")
        result = {}
        if CO2METER_HUM in self._values:
            result = {'humidity': (self._values[CO2METER_HUM]/100.0)}
        return result


    def get_data(self):
        result = {}
        result.update(self.get_co2())
        result.update(self.get_temperature())
        result.update(self.get_humidity())

        return result
